Gems/Atom/RPI/Code/Source/RPI.Public/Image/StreamingImageController.cpp

```diff
@@ -157,25 +205,191 @@ namespace AZ
+            return aznumeric_cast<uint16_t>(targetMip);
+        }
+
+        bool StreamingImageController::IsMemoryLow() const
+        {
+            // return true if m_lastLowMemory is not zero
+            return m_lastLowMemory;
+        }
+
+        bool StreamingImageController::EvictUnusedMips(StreamingImage* image)
+        {
+            uint16_t targetMip = GetImageTargetMip(image);
+            size_t targetMipChain = image->m_imageAsset->GetMipChainIndex(targetMip);
+            if (image->m_state.m_streamingTarget < targetMipChain)
+            {
+                RHI::ResultCode success = image->TrimToMipChainLevel(targetMipChain);
+            
+                StreamingDebugOutput("StreamingImageController", "Image [%s] mip level was evicted to %d\n",
+                    image->GetRHIImage()->GetName().GetCStr(), image->GetResidentMipLevel());
+
+                return success == RHI::ResultCode::Success;
+            }
+            return true;
+        }
+
+        void StreamingImageController::EvictUnusedMips()
+        {
+            AZStd::lock_guard<AZStd::mutex> lock(m_imageListAccessMutex);
+            if (m_streamableImages.size() == 0)
+            {
+                return; 
+            }
+
+            AZStd::list<StreamingImage*> evictedImages;
+
+            auto itr = m_streamableImages.begin();
+            while (itr != m_streamableImages.end())
+            {
+                StreamingImage* image = *itr;
+                if (EvictUnusedMips(image))
+                {
+                    itr = m_streamableImages.erase(itr);
+                    evictedImages.push_back(image);
+                }
+                else
+                {
+                    itr++;
+                }
+            }
+
+            for (auto image : evictedImages)
+            {
+                UpdateImagePriority(image);
+                m_streamableImages.insert(image);
+            }
+        }
+
+        bool StreamingImageController::OnHandleLowMemory()
+        {
+            StreamingDebugOutput("StreamingImageController", "Handle low memory\n");
+
+            // Evict some mips
+            bool evicted = EvictOneMipChain();
+
+            // Save the memory 
+            m_lastLowMemory = GetPoolMemoryUsage();
+
+            return evicted;
+        }
+
+        size_t StreamingImageController::GetPoolMemoryUsage()
+        {
+            size_t totalResident = m_pool->GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device).m_totalResidentInBytes.load();
+            return totalResident;
+        }
+                
+        size_t StreamingImageController::GetPoolFreeMemory()
+        {
+            const RHI::HeapMemoryUsage& memoryUsage = m_pool->GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device);
+            if (memoryUsage.m_budgetInBytes == 0)
+            {
+                return 4ul*1024*1024*1024;
```

Comment:
Does this hardcoded default value come from somewhere?

Reply:
It turns out this function is not used.